services/tns_submission_queue/tns_submission_queue.py

```diff
@@ -52,6 +52,618 @@ class TNSReportError(Exception):
+    ----------
+    submission_request : `~skyportal.models.TNSRobotSubmission`
+        The submission request.
+    source : `~skyportal.models.Source`
+        The source to submit.
+    tnsrobot : `~skyportal.models.TNSRobot`
+        The TNSRobot to submit with.
+    session : `~sqlalchemy.orm.Session`
+        The database session to use.
+
+    Returns
+    -------
+    reporters : str
+        The reporters string.
+    """
+    tnsrobot_id = submission_request.tnsrobot_id
+    user_id = submission_request.user_id
+    obj_id = submission_request.obj_id
+    if submission_request.custom_reporting_string not in [None, ""]:
+        reporters = submission_request.custom_reporting_string
+    else:
+        # if the user that first ever saved this source to a group that have access to this
+        # tnsrobot is not the user that is submitting, we make them the first author
+        # and the submitter the second author
+        author_ids = []
+        if (
+            submission_request.auto_submission
+            and source.saved_by_id is not None
+            and source.saved_by_id != user_id
+        ):
+            author_ids.append(source.saved_by_id)
+        author_ids.append(user_id)
+
+        coauthor_ids = [coauthor.user_id for coauthor in tnsrobot.coauthors]
+        for coauthor_id in coauthor_ids:
+            if coauthor_id not in author_ids:
+                author_ids.append(coauthor_id)
+
+        authors = (
+            session.scalars(sa.select(User).where(User.id.in_(author_ids)))
+            .unique()
+            .all()
+        )
+        # reorder the authors to match the order of the author_ids
+        authors = sorted(authors, key=lambda author: author_ids.index(author.id))
+
+        if len(authors) == 0:
+            raise TNSReportError(
+                f'No authors found for tnsrobot {tnsrobot_id} and source {obj_id}, cannot report to TNS.'
+            )
+        # if any of the users are missing an affiliation, we don't submit
+        authors_with_missing_affiliations = [
+            author for author in authors if len(author.affiliations) == 0
+        ]
+        if len(authors_with_missing_affiliations) > 0:
+            raise TNSReportError(
+                f'One or more authors are missing an affiliation: {", ".join([author.username for author in authors_with_missing_affiliations])}, cannot report {obj_id} to TNS.'
+            )
+
+        reporters = ', '.join(
+            [
+                f'{author.first_name} {author.last_name} ({author.affiliations[0]})'
+                for author in authors
+            ]
+        )
+        if tnsrobot.acknowledgments in [None, ""]:
+            raise TNSReportError(
+                f'No acknowledgments found for tnsrobot {tnsrobot_id}, cannot report {obj_id} to TNS.'
+            )
+
+        reporters += f' {str(tnsrobot.acknowledgments).strip()}'
+
+    return reporters
+
+
+def process_submission_request(submission_request, session):
```

Comment:
This function seems to still do generation + submission of the report; worth separating?

Reply:
Good point, will do!